vlasovsolver/gpu_trans_map_amr.cpp

```diff
@@ -130,168 +134,177 @@ __global__ void __launch_bounds__(WID3) translation_kernel(
+                     if (ti==0) {
+                        printf("Error in translation: trying to access LID %ul but sizes are vmesh %ul VBC %ul\n",blockLID,meshSize,VBCSize);
+                     }
+                  }
+                  #endif
                   if (ti==0) {
-                     printf("Error in translation: trying to access LID %ul but sizes are vmesh %ul VBC %ul\n",blockLID,meshSize,VBCSize);
+                     pencilBlockData[pencilBlockDataOffset + start + celli] = cellContainer->getData(blockLID);
+                     nonEmptyBlocks++;
                   }
+                  // Valid block, store values in contiguous data
+                  thisPencilOrderedSource[celli * WID3 + ti]
+                     = (cellContainer->getData(blockLID))[ti];
+               } else {
+                  if (ti==0) {
+                     pencilBlockData[pencilBlockDataOffset + start + celli] = NULL;
+                  }
+                  // Non-existing block, push in zeroes
+                  thisPencilOrderedSource[celli * WID3 + ti] = (Realf)(0.0);
                }
-               #endif
-               if (ti==0) {
-                  pencilBlockData[pencilBlockDataOffset + start + celli] = cellContainer->getData(blockLID);
-                  nonEmptyBlocks++;
-               }
-               // Valid block, store values in contiguous data
-               thisPencilOrderedSource[celli * WID3 + ti]
-                  = (cellContainer->getData(blockLID))[ti];
-            } else {
-               if (ti==0) {
-                  pencilBlockData[pencilBlockDataOffset + start + celli] = NULL;
-               }
-               // Non-existing block, push in zeroes
-               thisPencilOrderedSource[celli * WID3 + ti] = (Realf)(0.0);
+               __syncthreads();
+            } // End loop over this pencil
+            if (ti==0) {
+               pencilBlocksCount[pencilBlocksCountOffset + pencili] = nonEmptyBlocks;
             }
             __syncthreads();
-         } // End loop over this pencil
-         if (ti==0) {
-            pencilBlocksCount[pencilBlocksCountOffset + pencili] = nonEmptyBlocks;
-         }
-         __syncthreads();
-      } // end loop over all pencils
+         } // end loop over pencils
 
-      __syncthreads();
-      // Now we reset target blocks
-      for (uint celli=0; celli<sumOfLengths; celli++) {
-         if (pencilRatios[celli] != 0) {
-            // Is a target cell, needs to be reset
-            if (pencilBlockData[pencilBlockDataOffset + celli]) {
-               (pencilBlockData[pencilBlockDataOffset + celli])[ti] = (Realf)(0.0);
-            }
-         }
-      } // end loop over all cells
-
-      __syncthreads();
+         __syncthreads();
+         // Now we reset target blocks
+         for (uint pencilIndex = 0; pencilIndex < dev_binSize[nBin]; pencilIndex++) {
+            const uint pencili = dev_pencilsInBin[dev_binStart[nBin]+pencilIndex];
```

Comment:
As this section now changed from just clearing all target cells once to looping over all pencils and going through their target cells, I want to double-check the logic and counter use here before merging.